Terminal.Gui/Core/View.cs

```diff
@@ -568,17 +568,19 @@ void SetTabIndex ()
 				}
 
 				width = value;
-				if (autoSize && value.Anchor (0) != TextFormatter.Size.Width
-					- (TextFormatter.IsHorizontalDirection (TextDirection)
-					&& TextFormatter.Text.Contains (HotKeySpecifier) ? 1 : 0)) {
+
+				var isValidNewAutSize = autoSize ? IsValidAutoSizeWidth (width) : false;
+
+				if (autoSize && !isValidNewAutSize) {
+					TextFormatter.AutoSize = false;
```

Comment:
I agree with you, it will allow to users pay more attention of that and learning more by throwing an exception. Thanks.

Reply:
Just for reference the `AutoSize` in winforms says

> When using [AutoSize](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.form.autosize?view=windowsdesktop-6.0), [...] the current value of the [Size](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.control.size?view=windowsdesktop-6.0) property is ignored.

If we are throwing then we should make sure we are not creating a breaking change.

Maybe I'm not fully following but if before you could do 
```
myView.AutoSize = true;
myView.Width = 10;
```
Then it shouldn't suddenly start throwing Exception.  Especially if it doesn't throw when you do:

```
myView.Width = 10;
myView.AutoSize = true;
```

Reply:
It only throw the exception when the `IsInitialized` is true,

Reply:
Ok, that seems more fair.  In that case I think it should be [InvalidOperationException](https://docs.microsoft.com/en-us/dotnet/api/system.invalidoperationexception?view=net-6.0) with a message like "Width cannot be modified when when AutoSize is enabled"

Reply:
I already did like this:
`throw new InvalidOperationException ("Must set AutoSize to false before set the Width.");`

If not right please say.

Reply:
> Maybe I'm not fully following but if before you could do
> 
> ```
> myView.AutoSize = true;
> myView.Width = 10;
> ```
> 
> Then it shouldn't suddenly start throwing Exception. Especially if it doesn't throw when you do:
> 
> ```
> myView.Width = 10;
> myView.AutoSize = true;
> ```

@tznind setting the width and the height with absolute value that fit the auto size feature, it will stay true. But if you setting with values that doesn't fit the auto size when the `IsInitialized` is processed it set to false So, expressions like below that are set before the `IsInitialized` is processed will not throw even if don't fit the auto size. I have no choice to avoid to silent set to false on the `IsInitialized`, if it don't fit the auto size, otherwise it will throw the exception. I call your attention that the order which the object initializers are called are essential to avoid wrong values. Before by silent set to false  this wasn't an issue, so there must be coherence on how to order the objects. If you know how to bypass the order let me know, please.

`var myView = new View("Hi") { Width = 10,  AutoSize = true }; ` AutoSize will be true and the width will be 2.

`var myView = new View("Hi") { AutoSize = true, Width = 10 }; ` AutoSize will be false and width will be 10.

`var myView = new View("Hi") { Width = Dim.Fill(),  AutoSize = true };` AutoSize will be false and the width will be the parent width.

`Button` and `Checkbox` has the AutoSize enabled because the absolute values fit the auto size values.

Reply:
```
var myView = new View("Hi") { Width = 10, AutoSize = true };
```
To me this reads as 'set size to 10 but auto resize it if the text grows above that.  It is a bit annoying that the order of initializer matters.  Is there no way to store all the values then apply them at Initialize time in a consistent order?

Reply:
I also would like to know. It must be consistent at design time too. It's a good idea, considering a minimum size for auto size.

Reply:
But that minimum value will not be preserved after it grows, right?

Reply:
Well this is one more difficulty level. I don't know if it worth.

Reply:
> ```
> var myView = new View("Hi") { Width = 10, AutoSize = true };
> ```
> 
> 
>     
>   
> 
> To me this reads as 'set size to 10 but auto resize it if the text grows above that'.

@tznind, we must clarify about `AutoSize` does. I don't agree that the behavior you mentioned should be handle by the `View` but through a derived class, like `Button` and `Checkbox` do. Otherwise it will throw an exception if you decide at runtime to change the Width=20 with the text kept as "Hi" and auto size enabled. To do this we must first disable the auto size. To kept a minimum width a derived class must implement that property and deal with that. But perhaps I'm wrong about that.
`Dim` types other than `DimAbsolute` will always be used all the respective area and auto size must be disabled.
`DimAbsolute`can handled with auto size enable if his size fit the TextFormatter.Text, otherwise it must be disabled.
The `View` class can handle generic behaviors and not with special specific behaviors.
Since you are developing the great repo `TerminalGuiDesigner` you probably have more opinion about this.

Reply:
I end up to use the @tznind suggestion. If the user set the view `Width` equal to 10 and the `Text` length is smaller than 10, then the `Width` of 10 will be used. If `AutoSize` is true and the text grows above 10 the size will be adjust to the new auto-size length but the `Width` of 10 will always be kept. If the user doesn't set the view `Width`, so is null and the `Text` length is equal to 10, then the `Width` will be set to 10 in the `OnAdded` method. If `AutoSize` is true and the text grows above 10 the size will be adjust to the new auto-size length but the `Width` of 10 will always be kept. It's vital to know at design time if was the user who set the `Pos (X,Y)` and `Dim (Width,Height)` at object initializers. It can't be in the `Initialize` method of the base class nor on the `Initialize` method of the derived class. Object initializers I meant with objects that are set after the constructor execution, like this:
```cs
var view = new View() {
    TextDirection = TextDirection.TopBottom_LeftRight,
    AutoSize = true
}
```
If the `Pos/Dim` were set on the constructor initialization without the user action, with a text length equal to 10 the `Width` will also be sets. When the `TextDirection` is set with the old code the `Width` will keep 10 and the `Height` will be 10. This is wrong because the user didn't set the `Width` equal to 10. With the new code, the `Width` will be equal to auto-size length and the `Height` will be equal to 10. I wanted something that giving the same object state at design time and runtime. The `IsInitialized` only occur on `Application.Begin`, a little later and the `OnAdded` method occur early. The better solution will be some magical event will be fired on the base class when all the objects initializers are done on the derived class. But I think there is nothing that can handle that. The only care we must have is by views that hold views child's where the `Pos/Dim` aren't yet set, like the `Border` and `PanelView`, but they was fixed.